catroid/src/main/java/org/catrobat/catroid/web/WebConnection.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Catroid: An on-device visual programming system for Android devices
+ * Copyright (C) 2010-2018 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.web;
+
+import com.google.common.collect.ImmutableList;
+import com.squareup.okhttp.Call;
+import com.squareup.okhttp.Callback;
+import com.squareup.okhttp.ConnectionSpec;
+import com.squareup.okhttp.OkHttpClient;
+import com.squareup.okhttp.Request;
+import com.squareup.okhttp.Response;
+
+import java.io.IOException;
+import java.lang.ref.WeakReference;
+import java.net.MalformedURLException;
+import java.net.URL;
+
+public class WebConnection {
+	private final OkHttpClient okHttpClient;
+	private static final String USER_AGENT = "Mozilla/5.0 (compatible; Catrobatbot/1.0; +https://catrob.at/bot)";
+
+	private Call call;
+	private String url;
+	private WeakReference<WebRequestListener> weakListenerReference;
+
+	public WebConnection() {
+		okHttpClient = new OkHttpClient();
```

Comment:
okhttpclient is quite a heavy object, and it isnt meant to be initialized for every single call...

so this would be best to be moved over to the webConnectionHolder.

also after looking at the okhttpClient i found that this supports stuff like limiting amount of concurrent requests, etc. so probably worth inspecting if the whole webConncectionHolder could be refactored into just wrapping a single okHttpClient and doing the rest via configuration on OkHttpClient.

Reply:
I now handle the timeout via the okHttpClient, which works fine.
But limiting the amount of connection doesn't work with this user story.
Every connection, which is added after reaching the limit, is put to halt until a running connaction has finished.
The user story explicitly says to abort connections over the limit, and I failed to find a way to configure this behavior.

Please note, that Catrobat uses okHttp 2.3.0, and most google searches find documentaion on version 3.0.0 and later.

Reply:
since version 2.3.0 is now more than 4(!) years old, an update maybe overdue? and theres the possibility that this PR could be simplified using a newer version. so please investigate if updating is an option or if this causes a lot of problems with other server calls.
also research if okhttp call provides any information if a call is enqueued, either checking through dispatcher or if call itself can provide this information. if so, one could easily cancel enqueued calls. (those would be the ones when above max calls.)